iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/ExecutableNotOwnedByRootCheck.java

```diff
@@ -0,0 +1,127 @@
+
+  private static final String MESSAGE = "Make sure no write permissions are assigned to the executable.";
+  private static final String MESSAGE_SECONDARY_OTHER_EXEC = "Other executable file.";
+  private static final String MESSAGE_SECONDARY_CHOWN = "Sensitive file owner.";
+  private static final Set<String> SENSITIVE_FILE_EXTENSION = Set.of("exe", "py", "rb", "pl", "lua", "js", "lisp", "sh", "jar", "war",
+    "run", "bin", "bat", "ps1");
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(TransferInstruction.class, ExecutableNotOwnedByRootCheck::checkTransferInstruction);
+  }
+
+  private static void checkTransferInstruction(CheckContext ctx, TransferInstruction transferInstruction) {
+    Flag sensitiveChownFlag = getSensitiveChownFlag(transferInstruction);
+    if (sensitiveChownFlag != null) {
+      List<Argument> sensitiveFiles = getSensitiveFiles(transferInstruction.srcs());
+      Chmod chmod = getChmod(transferInstruction);
+
+      if (chmod == null) {
+        if (!sensitiveFiles.isEmpty()) {
+          reportIssue(ctx, sensitiveChownFlag, sensitiveFiles);
+        }
+      } else if (isChmodWriteSensitive(chmod) && (isChmodExecuteSensitive(chmod) || !sensitiveFiles.isEmpty())) {
+        reportIssue(ctx, sensitiveChownFlag, sensitiveFiles);
+      }
+    }
+  }
+
+  private static void reportIssue(CheckContext ctx, Flag sensitiveChownFlag, List<Argument> sensitiveFiles) {
+    if (sensitiveFiles.isEmpty()) {
+      ctx.reportIssue(sensitiveChownFlag, MESSAGE);
+    } else {
+      HasTextRange primaryLocation = sensitiveFiles.get(0);
+      List<SecondaryLocation> secondaryLocations = new ArrayList<>();
+      for (Argument otherExecutable : sensitiveFiles.subList(1, sensitiveFiles.size())) {
+        secondaryLocations.add(new SecondaryLocation(otherExecutable, MESSAGE_SECONDARY_OTHER_EXEC));
+      }
+      secondaryLocations.add(new SecondaryLocation(sensitiveChownFlag, MESSAGE_SECONDARY_CHOWN));
+      ctx.reportIssue(primaryLocation, MESSAGE, secondaryLocations);
+    }
+  }
+
+  @CheckForNull
+  private static Flag getSensitiveChownFlag(TransferInstruction transferInstruction) {
+    return transferInstruction.options().stream()
+      .filter(f -> f.name().equals("chown"))
+      .filter(ExecutableNotOwnedByRootCheck::isSensitiveUser)
+      .findFirst().orElse(null);
+  }
+
+  private static boolean isSensitiveUser(Flag chownFlag) {
+    ArgumentResolution resolvedArgArgument = ArgumentResolution.of(chownFlag.value());
+    return resolvedArgArgument.is(ArgumentResolution.Status.RESOLVED) && !"root".equals(resolvedArgArgument.value());
+  }
+
+  @CheckForNull
+  private static Chmod getChmod(TransferInstruction transferInstruction) {
+    return transferInstruction.options().stream()
+      .filter(f -> f.name().equals("chmod"))
+      .map(f -> ArgumentResolution.of(f.value()))
+      .filter(argResolved -> argResolved.is(ArgumentResolution.Status.RESOLVED))
```

Comment:
The call `argResolved.is(ArgumentResolution.Status.RESOLVED)` looks a little bit strange because the internal implementation detail is exposed outside the class. 
Consider adding methods `ArgumentResolution.isResolved()`, `ArgumentResolution.isUnresolved()` & `ArgumentResolution.isEmpty()` - then client code will be much better

Reply:
Yeah, sounds good to me.